### PR TITLE
Add Accumulator3D

### PR DESCRIPTION
**Pull request by SergeyTsimfer**
This PR improves on @a-arefina work by allowing us to accumulate predicted crops either in memory or on disk, while also adding more aggregations and QoL changes to placeholder creation.

I will also include the following changes as soon as possible:

- [x] change `HorizonController` to use new accumulate logic
- [ ] add non-zero `prefetch` value for inference of `HorizonController`
- [x] correct `SeismicCubeset.show_slices` with `mask_rebatch`
- [x] augment `SeismicGeometry` repr with the size of original `SEG-Y` cube and area of the processed field

### CODE REVIEW

**Comment by SergeyTsimfer** [comment]
As of now, `HorizonController.inference` correctly works with non-zero prefetch, but leaks memory

File: seismiqb/src/utility_classes.py [hunk 1]
```diff
@@ -224,6 +226,239 @@ def get(self, final=False):
...
+            self.options = {**kwargs}
+        self.type = 'hdf5' if path is not None else 'numpy'
+
+        self.names = []
+        self.aggregated = False
+
+        # Create underlying storages
+        self._init()
+
+    def _init(self):
```
Comment (a-arefina) [inline seismiqb/src/utility_classes.py:293]: why did you remove __init__ methods in child classes? You could pass specialized arguments there, instead of mixing them all in base classes __init__ or parsing kwargs

**Review by a-arefina — COMMENTED** [review]
still reflecting on placeholders, but here are some comments

File: seismiqb/src/utility_classes.py [hunk 2]
```diff
@@ -451,11 +451,15 @@ def _aggregate(self):
...
-                self.data[i] = np.pow(self.data[i], 1/counts)
+
+                counts = counts.astype(np.float32)
+                counts **= -1
+                self.data[i] **= counts
 
         elif self.type == 'numpy':
             self.counts[self.counts == 0] = 1
-            self.data = np.pow(self.data, 1/self.counts)
+            self.counts **= -1
```
Comment (a-arefina) [inline seismiqb/src/utility_classes.py:461]: add 
counts = counts.astype(np.float32)

File: seismiqb/src/utility_classes.py [hunk 3]
```diff
@@ -224,6 +226,247 @@ def get(self, final=False):
...
+        _ = crop, location
+        raise NotImplementedError
+
+    def aggregate(self):
+        """ Finalize underlying storages to create required aggregation. """
+        if self.aggregated:
+            raise RuntimeError('All data in the container has already been cleared!')
+        self._aggregate()
+
+        # Re-open the HDF5 file to optionally repack it
```
Comment (a-arefina) [inline seismiqb/src/utility_classes.py:339]: plz update the comment to describe why we need to reopen the file

File: seismiqb/src/utility_classes.py [hunk 4]
```diff
@@ -224,6 +226,247 @@ def get(self, final=False):
...
+        After the initialization, we keep the file handle in `w-` mode during the update phase.
+        After aggregation, we re-open the file to automatically repack it in `r` mode.
+    kwargs : dict
+        Other parameters are passed to HDF5 dataset creation.
+    """
+    def __init__(self, shape=None, origin=None, dtype=np.float32, transform=None, path=None, **kwargs):
+        # Dimensionality and location
+        self.shape = shape
+        self.origin = origin
+
```
Comment (a-arefina) [inline seismiqb/src/utility_classes.py:270]: maybe add self.data = None to show that this field must exist?

**Review by a-arefina — APPROVED** [review]
please see two minor comment and off we go!
